### DISCUSSION TITLE
IB newbie trying to figure out the best way to architect an app

### DISCUSSION THREAD

**Discussion by Graeme22** [ib-api-reloaded/ib_async#30] [Q&A] [answered]
Greetings!

I'm currently developing a copy trading bot for a client. We have support for Tastytrade and would like to add IB as well.

Now, as I'm reading up on the IB APIs and different options, I'm trying to figure out the best way to design the app. Originally I was looking at the web API, but it seems like any sessions created will be invalidated by a login elsewhere which is problematic as ideally the users could be logged in to the desktop platform and see trades coming in. Am I correct in assuming this wouldn't work?

Assuming that's the case, we'd then need to consider using the TWS API. However, we'd be placing trades for around 300 users at once. Would we be able to route all of those trades through the same gateway without running into rate limits or other issues? I wasn't clear on the rate limit being per gateway or per user.

Appreciate any help from the IB experts here. Thanks in advance!

Graeme

**Comment by mattsta** [comment, marked as answer]
> any sessions created will be invalidated by a login elsewhere which is problematic as ideally the users could be logged in to the desktop platform and see trades coming in. Am I correct in assuming this wouldn't work?

Yes, the single-account login is just an IBKR limitation. It's how they protect their market data agreements by only allowing one connection per account (sometimes you can play tricks with alternate users or [sub accounts](https://www.ibkrguides.com/clientportal/createnewlinkedaccount.htm) all attached to the same parent account).

If you want users to see their account details while you are connected to their IBKR account, you would need to create a second custom dashboard for your users based on the internal trade details (either cached activity or _your dashboard_ calls the IBKR API calls to fetch live data itself, etc).

Other API brokers like [Tradier are 100% REST based](https://documentation.tradier.com/brokerage-api) and don't require a gateway (yet, other API brokers also require a [local gateway](https://openapi.moomoo.com/moomoo-api-doc/en/)).

> we'd be placing trades for around 300 users at once.

The IBKR gateway runs _per account_, so it all depends on how your accounts are setup.

If each of your accounts are individual accounts you are running _for_ other users, you would need 300 Java gateway processes (and they all require manual 2FA after user/pass login unless you ask IBKR to disable it on your account).

If you have an advisor account and all the sub-accounts are just managed clients of your primary account, then you would only need one gateway connection and you would just manage all the sub-account details yourself about which accounts get which trades executed.

As for rate limits, the official docs just say:

> Aside from the TWS API's inherent limitation of 50 messages per second implying a maximum of 50 orders per second being sent to the TWS, there are no further API-only limitations.

50 messages per second is 3,000 messages per minute, and it seems like enough for standard trading operations.

> Additionally, please note IB allows up to 20 active orders per contract per side per account.

Also IBKR gets upset if you submit many orders which aren't executing. Their stated limit is you should remain under than 20 order calls (including modifications or cancels) per 1 executed order (so, just consider you start the day with 20 free order commands, then every executed order gives you an additional 20 order commands). Basically they don't want people submitting $0.01 limit buy orders on 1,000 symbols every day, so they expect orders to be "meaningful" and likely to execute.

**Reply by Graeme22** [reply]
This is great info, thanks!

The gateway being per account is bad news. In that case we may just have to accept the limitation and use the REST API, since these aren't managed accounts, just individual accounts we're automating trades for. We'd probably just have to create a fresh session before placing each trade to be safe. Or consider supporting a different broker, though IBKR covers a lot of countries. Rate limits wouldn't be an issue, we're just placing one trade per user at a time.

Let me know if I'm not on the right track, and thanks for the help.

**Reply by mattsta** [reply]
Yeah, sounds about right. There's no clean way to automate other people's accounts with IBKR unless you are their managing broker and they are clients of yours (but then you can do other things like automate taking extra commission fees or percentages of funds too).

I think the cleanest broker to fit your use case would be Tradier since they support oauth 3rd party manager vs. client relationships so you never need the client username/password, but IBKR doesn't have a way to run trades on an account without full 2FA account access to the account itself.

Good luck!

**Reply by Graeme22** [reply]
We'd really like to support IB since it has so many users around the globe, but it does seem like it'd be quite tricky.

Quick question about the one session per account limitation: is that per trading account? Or per user account?

Thanks.

**Reply by mattsta** [reply]
There are various docs around it ("sub accounts" vs "linked accounts" vs "multiple usernames per account") but it's best to just play with the system to figure out how the features work too:

- https://www.ibkrguides.com/clientportal/uar/addingauser.htm
- https://www.ibkrguides.com/clientportal/uar/addingusernamestoauser.htm
- https://ibkrcampus.com/trading-lessons/creating-users-and-user-roles/
- https://www.interactivebrokers.com/en/trading/linked-accounts.php

**Reply by Graeme22** [reply]
Thanks, I'll take a look.
